Model/Api/OAuthRedirect.php

```diff
@@ -35,37 +46,198 @@ class OAuthRedirect implements OAuthRedirectInterface
+        StoreManagerInterface $storeManager,
+        CustomerSession $customerSession,
+        CustomerInterfaceFactory $customerInterfaceFactory,
+        CustomerFactory $customerFactory,
+        Url $url,
+        Bugsnag $bugsnag
     ) {
         $this->response = $response;
         $this->deciderHelper = $deciderHelper;
+        $this->ssoHelper = $ssoHelper;
+        $this->externalCustomerEntityRepository = $externalCustomerEntityRepository;
+        $this->customerRepository = $customerRepository;
+        $this->storeManager = $storeManager;
+        $this->customerSession = $customerSession;
+        $this->customerInterfaceFactory = $customerInterfaceFactory;
+        $this->customerFactory = $customerFactory;
+        $this->url = $url;
+        $this->bugsnag = $bugsnag;
     }
 
     /**
      * Login with Bolt SSO and redirect
      *
      * @api
      *
-     * @param string $authorization_code
+     * @param string $code
+     * @param string $scope
+     * @param string $state
      *
      * @return void
      *
      * @throws NoSuchEntityException
+     * @throws WebapiException
      */
-    public function login($authorization_code = '')
+    public function login($code = '', $scope = '', $state = '')
     {
         if (!$this->deciderHelper->isBoltSSOEnabled()) {
             throw new NoSuchEntityException(__('Request does not match any route.'));
         }
 
-        $this->response->setHeader('Content-Type', 'application/json');
-        $this->response->setHttpResponseCode(500);
-        $this->response->setBody(json_encode(['message' => 'Not implemented']));
-        $this->response->sendResponse();
+        if ($code === '' || $scope === '' || $state === '') {
+            throw new WebapiException(__('Bad Request'), 0, WebapiException::HTTP_BAD_REQUEST);
+        }
+
+        $oauthConfiguration = $this->ssoHelper->getOAuthConfiguration();
+        $clientID = $oauthConfiguration['clientID'];
+        $clientSecret = $oauthConfiguration['clientSecret'];
+        $boltPublicKey = $oauthConfiguration['boltPublicKey'];
+
+        $token = $this->ssoHelper->exchangeToken($code, $scope, $clientID, $clientSecret);
+        if ($token === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $payload = $this->ssoHelper->parseAndValidateJWT($token, $clientID, $boltPublicKey);
+        if ($payload === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $websiteId = $this->storeManager->getStore()->getWebsiteId();
+        $storeId = $this->storeManager->getStore()->getId();
+
+        $externalCustomerEntity = null;
+        $customer = null;
+        try {
+            $externalCustomerEntity = $this->externalCustomerEntityRepository->getByExternalID($payload['sub']);
+        } catch (NoSuchEntityException $nsee) {
```

Comment:
Is there is any way to get rid of empty catch blocks? If not could you add a comment?

Reply:
will add comment